### PR TITLE
[BEAM-1047] Update dataflow runner code to use DataflowClient wrapper.

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by peihe** [comment]
R: @bjchambers

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import com.google.api.services.dataflow.model.ListJobsResponse;
+import com.google.api.services.dataflow.model.ReportWorkItemStatusRequest;
+import com.google.api.services.dataflow.model.ReportWorkItemStatusResponse;
+import java.io.IOException;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions;
+
+/**
+ * Client library for {@link Dataflow}.
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: Be more descriptive -- this is already wrapping a client library. Maybe:

"Wrapper around the generated {@link Dataflow} client to provide common functionality."

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  private DataflowClient(Dataflow dataflow, DataflowPipelineOptions options) {
+    this.dataflow = checkNotNull(dataflow, "dataflow");
+    this.options = checkNotNull(options, "options");
+  }
+
+  /**
+   * Creates the Dataflow {@link Job}.
+   */
+  public Job createJob(@Nonnull Job job) throws IOException {
+    Jobs.Create jobsCreate = dataflow.projects().jobs().create(options.getProject(), job);
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: I would suggest making the constructor take the parts of the options that are necessary -- getProject, etc. and storing those in fields. That will make it clearer what parameters the `DataflowClient` actually uses.

Also, you already do some of this in the `create` call (pulling out the `dataflowClient`), so this makes things more consistent (and likely easier to test).

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 3]
```diff
@@ -146,7 +144,7 @@ public String getJobId() {
    * Get the project this job exists in.
    */
   public String getProjectId() {
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: Where is this being used? Is it necessary?

Also, consider extracting just the project ID from the options. Using underlying options has two potential problems. It makes it unclear which fields you're using and it makes it possible that the value changes during execution if the options are mutated. Better to just hold on to the string directly.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java [hunk 4]
```diff
@@ -279,8 +282,7 @@ boolean atMaxWatermark(DataflowPipelineJob job, JobMetrics metrics) {
   JobMetrics getJobMetrics(DataflowPipelineJob job) {
     JobMetrics metrics = null;
     try {
-      metrics = options.getDataflowClient().projects().jobs()
-          .getMetrics(job.getProjectId(), job.getJobId()).execute();
+      metrics = dataflowClient.getJobMetrics(job.getJobId());
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: Would it make sense to allow construction of a dataflowClient that was bound to a job? Seems like the most common case...

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/MonitoringUtil.java [hunk 5]
```diff
@@ -67,8 +67,8 @@
   private static final String JOB_MESSAGE_DETAILED = "JOB_MESSAGE_DETAILED";
   private static final String JOB_MESSAGE_DEBUG = "JOB_MESSAGE_DEBUG";
 
-  private String projectId;
-  private Messages messagesClient;
+  private final String projectId;
+  private final DataflowClient dataflowClient;
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/MonitoringUtil.java]: If the `projectId` was available from the `dataflowClient` then we could get rid of the duplication of these fields in multiple places.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 2 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/MonitoringUtil.java (hunk 5 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/MonitoringUtil.java]: removed

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java (hunk 4 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: Not all methods take jobId.
I feel it is not clear in the code to implicitly bound to an Nullable jobId.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java (hunk 3 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: It is used in this class's other functions.
But, it is public, and could be useful for users. So, I am not confident for deleting it.

I agree with the mutation issue. However, dataflowOptions is also used for MonitoringUtil.getGcloudCancelCommand(). The current change will make the code consistent at least.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 1 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  private DataflowClient(Dataflow dataflow, String projectId) {
+    this.dataflow = checkNotNull(dataflow, "dataflow");
+    this.projectId = checkNotNull(projectId, "options");
+  }
+
+  /**
+   * Creates the Dataflow {@link Job}.
+   */
+  public Job createJob(@Nonnull Job job) throws IOException {
+    Jobs.Create jobsCreate = dataflow.projects().jobs().create(projectId, job);
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: Should we add a Preconditions.checkNotNull on the job?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        .list(projectId)
+        .setPageToken(pageToken);
+    return jobsList.execute();
+  }
+
+  /**
+   * Updates the Dataflow {@link Job} with the given {@code jobId}.
+   */
+  public Job updateJob(@Nonnull String jobId, @Nonnull Job content) throws IOException {
+    Jobs.Update jobsUpdate = dataflow.projects().jobs()
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: Preconditions.checkNotNull?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        .update(projectId, jobId, content);
+    return jobsUpdate.execute();
+  }
+
+  /**
+   * Gets the Dataflow {@link Job} with the given {@code jobId}.
+   */
+  public Job getJob(@Nonnull String jobId) throws IOException {
+    Jobs.Get jobsGet = dataflow.projects().jobs()
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: .

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        .get(projectId, jobId);
+    return jobsGet.execute();
+  }
+
+  /**
+   * Gets the {@link JobMetrics} with the given {@code jobId}.
+   */
+  public JobMetrics getJobMetrics(@Nonnull String jobId) throws IOException {
+    Jobs.GetMetrics jobsGetMetrics = dataflow.projects().jobs()
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: .

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        .getMetrics(projectId, jobId);
+    return jobsGetMetrics.execute();
+  }
+
+  /**
+   * Lists job messages with the given {@code jobId}.
+   */
+  public ListJobMessagesResponse listJobMessages(
+      @Nonnull String jobId, @Nullable String pageToken) throws IOException {
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: .

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        .setPageToken(pageToken);
+    return jobMessagesList.execute();
+  }
+
+  /**
+   * Leases the work item for {@code jobId}.
+   */
+  public LeaseWorkItemResponse leaseWorkItem(
+      @Nonnull String jobId, @Nonnull LeaseWorkItemRequest request) throws IOException {
+    Jobs.WorkItems.Lease jobWorkItemsLease = dataflow.projects().jobs().workItems()
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: .

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        .lease(projectId, jobId, request);
+    return jobWorkItemsLease.execute();
+  }
+
+  /**
+   * Reports the status of the work item for {@code jobId}.
+   */
+  public ReportWorkItemStatusResponse reportWorkItemStatus(
+      @Nonnull String jobId, @Nonnull ReportWorkItemStatusRequest request) throws IOException {
+    Jobs.WorkItems.ReportStatus jobWorkItemsReportStatus = dataflow.projects().jobs().workItems()
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: .

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 6 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 7 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 8 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 9 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 10 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 11 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java (hunk 12 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowClient.java]: done

**Pull request closed without merging**